### PR TITLE
Rework the state layer of Discord4J

### PR DESCRIPTION
**Pull request by Alex1304**
**The description of this PR is updated according to the latest discussions about architectural changes happening in this PR and on the Discord server.**

## Description

* `discord4j-core` no longer depends on the old `stores-api` (Discord4J/Stores)
* In `discord4j-common` a new package `discord4j.common.store` was created. It contains a `Store` class which has 1 public method `execute(action)`. It is constructed via `Store#fromLayout(layout)`, `StoreLayout` being an interface that may be moved to a separate module above d4j-common later, it defines the minimal set of actions required by d4j as well as the possibility to add custom action mappings. A default `StoreLayout` implementation is proposed via the PR #799.
* Store actions that deal with gateway updates are fired directly in the `discord4j-gateway` module via the new class `DispatchStoreLayer`. It is independent from the gateway client itself, it allows to keep a store updated in projects that only depends on `discord4j-gateway`.
* `*DispatchHandlers` classes in `discord4j-core` module have been significantly simplified, now they no longer do any store operations, their sole purpose now is only to convert dispatches to core Event entities (and async request members for guild create). A third generic type parameter had to be added on `DispatchContext` to know the type of the old state. 
* `StateHolder` has been removed, all references to it were replaced by the new `Store` class. 

## Justification

The current state layer provided by Discord4J via the Discord4J/Stores project is showing its limitations, as the current interface only has two generic methods to save and find entities, but no way to send specific updates when a gateway update event is received. It currently has to retrieve the whole entity first, then update it locally, then re-insert it back to the Store which is extremely costly in some scenarios. This new API defines layouts to handle individual gateway events and find operations required by d4j, while opening the path for specific store implementations to handle custom queries.

### CODE REVIEW

**Comment by quanticc** [comment]
Thank you very much for creating this PR, as we're looking into reworking our state layer, this is an excellent start. We definitely need the feedback from the team to move forward but I think it's aligning pretty well with at least my current vision.

Enabling implementations to deal with gateway dispatch directly certainly makes them more complex, but the added flexibility allows each one to determine their own layout or structure in an effort to optimize commands and queries.

I'm identifying the style decisions around this rework that may impact this PR and also the stores one. I detail them here to guide further discussion among the team:

#### Architecture changes

Whether d4j-gateway needs to depend on the new Store API or stick to d4j-core. I'm still considering this, not 100% sold on it but it seems like a good idea if you want a gateway module that is also capable of keeping a store updated and thus not require importing core fully.

Another topic is the fact DispatchContext now must keep a generic type for the old state that might be optionally retrieved from the store. I would like the team to agree that `StateAwareDispatch::getOldState` is the preferred approach.

#### Command-oriented interface vs method-oriented interface

This PR and Discord4J/Stores#9 uses the former, calling them "Actions" which means we need to catalogue each and create those additional classes, as opposed to multiple methods in a single interface. Being command-oriented is a nice encapsulation of a change/query which means it can be decorated for batching, transactions, queuing.

I like this approach but implementors need to know each action from somewhere, because nothing will statically force them to implement them - which can be a good thing too for evolution.

#### Source location and naming

It's essential that Stores from 3.1.x work with this so we don't break those implementations. I think we need a new package for the new stuff but I don't have a clear option for now. Should it still be called "Store", for example and instead of "wip" package we could use `v2` or even a fantasy term indicating the new architecture. For example at one point we were intending to call the next-generation of stores "Catalyst".

About the naming of other classes, we'll see if they make sense and it's not relevant at this point. For example ReadAction/WriteAction vs Query/Command. StateAwareDispatch vs StatefulDispatch.

**Comment by Alex1304** [comment]
Ok so, until we can get more feedback from the rest of the team before we can consider making a default `Store` implementation under this new API, in the meantime I wanted to do more work towards the removal of `StateHolder`, to see if more problematics or blocking points would arise. I finally did it, now `StateHolder` is no more, but I have found two subjects that deserve discussion : shard invalidation and handling of incomplete list of entities.

## Shard invalidation

When I went to edit the code in `GatewayBootstrap`, at first I didn't really know what I was supposed to do with `ShardInvalidationStrategy` and the code that invalidates the stores on disconnect. In this new API, invalidating stores for a shard would be as simple as doing something like `store.execute(new InvalidateShardAction(shardIndex))`, then the implementor is free to interpret this action as they see fit. Then it's all about determining where to put this call. I think we can all agree that shard invalidation is supposed to happen in two cases:

* When logging out
* When a hard reconnect happens (non-resumable session)

The former is the easiest, when logging out a `GatewayStateChange` is dispatched, so in the `DispatchStoreLayer` I just had to intercept this dispatch with a condition `getState() == DISCONNECTING`. For the latter however, it is more complicated as currently there is no specific dispatch when a hard reconnect happens. The 3.1 code invalidates shards when a `RETRY_FAILED` happens twice in a row, but that's not accurate because the second failure is not necessarily caused by a non-resumable session. So my idea was to create a new dispatch called `GatewayHardReconnect`, that would be dispatched whenever the `GatewayConnection.State` switches to `RECONNECTING`, and that is also intercepted in the `DispatchStoreLayer`. I need some feedback on this as I don't know how comfortable you would be with adding new dispatches like that, but for sure this is now the wanted behavior.

## Handling of incomplete list of entities

Do you remember https://github.com/Discord4J/Discord4J/issues/708 and https://github.com/Discord4J/Discord4J/pull/713? This created a complete mess in the code of `StoreEntityRetriever`:
```java
@Override
public Flux<Member> getGuildMembers(Snowflake guildId) {
    return stateView.getGuildStore().find(guildId.asLong())
            .filterWhen(guild -> Mono.justOrEmpty(gateway.getGatewayResources().getIntents().toOptional())
                    .filter(intents -> intents.contains(Intent.GUILDS))
                    .flatMap(intents -> new Guild(gateway, guild).getVoiceStates().count())
                    .map(count -> guild.memberCount() - count == 1)
                    .defaultIfEmpty(true)) // Fix for https://github.com/Discord4J/Discord4J/issues/708
            .flatMapMany(guildData -> Flux.fromIterable(guildData.members())
                    .flatMap(memberId -> stateView.getMemberStore()
                            .find(LongLongTuple2.of(guildId.asLong(), Snowflake.asLong(memberId))))
                    .map(member -> new Member(gateway, member, guildId.asLong())))
            .switchIfEmpty(gateway.requestMembers(guildId));
}
```
Seriously what even is this? I have trouble to understand what this code does. The base issue is that under certain conditions, the list of members might be incomplete, because either it didn't yet receive the events to populate the member cache, or it invalidated the cache for whatever reason. I'm 99% sure reasoning like that with intents and voice state cache is buggy and doesn't fit all use cases.

So, when migrating this piece of code to use the new API, what solution did I bring? Well, a code snippet will explain it better than words I think:
```java
@Override
public Flux<Member> getGuildMembers(Snowflake guildId) {
    return store.execute(new GetGuildMembersAction(guildId.asLong())) // Mono<PossiblyIncompleteList<ChannelData>>
            .flatMapMany(list -> {
                if (list.isComplete()) {
                    return Flux.fromIterable(list)
                            .map(data -> new Member(gateway, data, guildId.asLong()));
                }
                return gateway.requestMembers(guildId);
            });
}
```
I created a new type called `PossiblyIncompleteList<E>` that is basically a `List<E>` + a `boolean` that gives a hint on whether the list is complete or incomplete. It only serves as a hint and does not provide any guarantee: `Store` implementations are not forced to compute it, but it would be encouraged by docs to allow clients to save some queries to Discord API to complete the list. I extended the use of this type to `getRoles()`, `getChannels()`, etc, even if it's less problematic for them it might be useful if for example the `Store` implementations uses automatic eviction of roles, channels, etc (e.g caffeine), it might return an incomplete list. Only exception is `getUsers()`, because well it is impossible to cache all users that ever exist on Discord, the list will always be incomplete. If you're wondering about `getGuilds()`, only the guilds the bot is in are relevant, so this type works.

In order to allow the store to know whether the list of members is complete or not, it needs to intercept request members actions and mark as complete if the request was the full member list (`query = ""` and `limit = 0`). I added `RequestMembersAction`, and I added code in `GaewayClientGroup` implementations to execute this action if the payload passed to `.unicast(...)` and `.multicast(...)` is of type `RequestGuildMembers`. I think that's less tedious than keeping track of everywhere a request members is initiated, however it implies that `GatewayClientGroup` is aware of the store instance and it implied a signature change to `ShardingStrategy#getGroupManager` to accept a `Store`.

On this topic I need feedback on:
* The relevance of `PossiblyIncompleteList` and the solution I brought for the `getGuildMembers()` stuff
* The `RequestMembersAction` and the signature change in sharding strategy

Now to come back on @quanticc 's previous comment, I 100% agree that `StatefulDispatch` is a much better name than `StateAwareDispatch`, it sounds less misleading, so I went ahead and changed it. For the packange name however I'm still not sure, but`v2` seems to be the only viable option we have for now. But I have another proposal also, since now the stores version does not necessarily match the d4j version, I think it would be safe to release this new API under the version 4.0.0 with only the new classes, and maintain the old classes on the 3.x branch for future potential bugfixes. For the naming of `Action` classes, I think it gives more consistency to suffix all of them with `Action`, query and commands would break that pattern. But well that's just my opinion.

**Comment by darichey** [comment]
These are great ideas, thank you for putting in the time here. I need to think on things some more, but here are some of my thoughts so far:

I believe this approach was born from the discussion we had in the D4J guild about creating a less-abstract store interface (i.e., one which is more aware of its purpose: storing state as it pertains to _Discord_). As you say in Discord4J/Stores#9, we don't need a generic CRUD interface. That said, I think it's interesting how we really ended up with something even more abstract! :smile: Seeing it now, that's not necessarily a bad thing, but I think I'm not convinced...

My biggest concern is the one @quanticc mentioned:
> I like this approach but implementors need to know each action from somewhere, because nothing will statically force them to implement them - which can be a good thing too for evolution.

As it stands, store implementors have no idea which actions are the "minimal set" they need to support everything D4J is going to ask of them. This is unfortunate considering that set _can_ be statically known (ctrl+f, "action"), but we have no way of communicating it to them. Maybe there is something to be had for evolution, but with the current location of Actions being in the store api, that minimal set can't even change without a version bump anyways. Even if the actions did live in `core`, we could add a new one, but all existing stores wouldn't respond to it. I think the method-based approach solves all of these problems. We should discuss the pros and cons further.

**Comment by Alex1304** [comment]
Well, honestly I think if there was a way to statically force implementors to implement actions, the main problem would be that backward compatibility would be broken whenever new actions are added. Having an interface with only default methods could be a compromise, but then we would lose in flexibility, every possible action would be in that single interface (it wouldn't really be possible to split/categorize them). 

With an Action interface approach, we can have a split between actions coming from gateway, actions to read data, etc. If in the future we add support for REST caching for example, another whole hierarchy of Action concrete classes could be created for that. But then, I understand the concern of not being able to determine the minimum set of Actions to implement, and also about the location of these Action classes. I agree that bumping stores version just because we forgot to add an action is kinda uhh. I would imagine that Actions could reside in a separate repository, which means for example that `discord4j-gateway` would depend on `actions-gateway`, and `discord4j-core` would depend on `actions-read` (that contains all GetXById actions and such).

But now that I think about it, it would be totally possible to have some sort of templates for implementors to use with this model. For now I only made a class `StoreActionHandlingHelper` that accepts single method references for one action (see the other PR for example of use), but this could easily be extended to accept templates with some predefined set of actions. Those templates would reside in the respective actions-x projects.

**Comment by darichey** [comment]
I'm a bit confused about the "backwards compatibility" aspect. There's a few options, none of which provide backwards compatibility, from what I can tell.

## Option A
Store actions (commands, methods, or otherwise) are defined by `stores-api`.

If a new version of D4J requires a new store action, it must be added in `stores-api` and the version is bumped. Any store implementation must upgrade to this new version to support the new action.

## Option B
Store actions are defined by `discord4j-core`.

If a new version of D4J requires a new store action, it is added in `discord4j-core`. Technically, this requires no version bumps anywhere. Practically, without a version bump, no store implementation can actually support the new action. So while binary compatibility is maintained, it's basically useless.

## Option C
Store actions are defined by some new libraries, `actions-gateway`, `actions-read`, etc..

If a new version of D4J requires a new store action, it is added to the appropriate `actions-*` and the version is bumped. Any store implementation must upgrade to this new version to support the new action.

It seems like Option A and Option C are basically equivalent from this point of view. Option B seems to be the only one that allows for binary compatibility, but, as I said, it doesn't seem to be a useful kind of compatibility here.

**Comment by Alex1304** [comment]
By "backwards compatibility" I was specifically referring to the design where you would have a bunch methods instead of Action objects. Adding a method to an interface will break backward compatibility, unless that method is `default`.

About your options, I don't think B is good. It would require implementations to import the whole `discord4j-core` module just so it can access the action classes, which might not always be wanted. It would also make it impossible to make the default store implementation reside in the current Stores repository (`stores-jdk` or `stores-something`), as releasing a new D4J version would require Stores to be released prior and vice-versa (circular dependency on release, same issue that prevents the use of Snowflake type in actions). Not to mention that the `discord4j-gateway` would need to access them, because it's where gateway-related actions are fired.
About A and C, personally I'm indifferent whether we go for one or the other. C is the solution that allows to import the least dependencies when implementing a store, but would require you to maintain an extra repository. That's a choice to make and I think that's up to you and the team.

In the meantime, I updated this PR to try to propose something regarding your concerns. I reworked `StoreActionHandlingHelper` (which is now called simply `ActionHandler`), now it allows to wire interface methods that the Store can implement those methods, which should solve the "impls don't know what to implement" issue. It would work as follows:

```java
public class MyStore implements Store, ReadActionHandler, GatewayActionHandler {

    private final ActionHandler actionHandler;

    public MyStore() {
        this.actionHandler = ActionHandler.withMappers(
                ReadActionHandler.mapper(this),
                GatewayActionHandler.mapper(this));
    }

    @Override
    public <R> Mono<R> execute(StoreAction<R> action) {
        return actionHandler.handle(action);
    }

    // Override methods from ReadActionHandler and GatewayActionHandler here
}
```

**Comment by darichey** [comment]
> By "backwards compatibility" I was specifically referring to the design where you would have a bunch of methods instead of Action objects. Adding a method to an interface will break backward compatibility, unless that method is default
Right, this is what I meant by the binary compatibility. But I think broader concerns about "compatibility" apply to either approach.

Is a store implementation that always returns empty for new actions _really_ compatible with those new actions in any useful sense? I personally don't think so.

**Comment by Alex1304** [comment]
I'm thinking about cases where a user would maintain his own Store implementation in his own project. If a new D4J and Stores versions is released with some new actions, that user would be forced to also update his implementation immediately or else it will get "abstract method not implemented" errors at runtime, while they may prefer upgrading D4J first and then upgrade his Store impl. That would be fine for example if the Minor version of D4J/Stores is bumped (e.g `3.2.5` -> `3.3.0`), because D4J's versionning convention allows breaking changes at that level, but it would be less okay if those new actions are added in a Patch version (e.g `3.2.0` -> `3.2.1`). So unless you can confirm me that you are going to completely ban the addition of new actions on patch versions, I would prefer going for the "default empty" approach which is safer.

**Comment by Alex1304** [comment]
I think I found what could be the worst possible scenario. If let's say the bot developer who depends on D4J also depends on a Store implementation that is maintained by a third-party (neither D4J team or the dev himself). If a new D4J and Stores version is released, the bot dev wouldn't be able to upgrade until the maintainer of the Stores impl has added support for the new actions. You can see now how problematic it would be.

Another alternative I can suggest is to log a warning instead of just returning empty. That way in this scenario the bot dev would be able to report the warning to the Store impl maintainer, and continue to use the current impl with the new D4J version.

For comparison it would work kinda like when Discord adds a new JSON type, D4J would just log a warning and wouldn't break completely. That warning could be reported as a new issue in this repository to further add support for it.

**Comment by darichey** [comment]
I think it makes most sense for a store to support a certain version of D4J. If store-impl supports D4J 3.x, and then D4J 3.(x+1) is released with some new store actions, then store-impl needs to update to support those actions. 

In the worst case scenario, Discord adds some new entity. D4J 3.(x+1) adds support for that entity, which requires many new store actions (which are added to the appropriate place according to which ever "Option"). However, even if a bot developer upgrades to this version, if they use an old store-impl, then virtually all of the new features are unsupported (getNewEntityById will never work). This version of store-impl does not support this version of D4J, regardless of whether it can compile and run alongside it or not.

I don't agree that your scenario is really so bad. As I laid about above, that store doesn't support that version of D4J. It's unlikely that the bot developer could upgrade versions of D4J and still have a working bot if their third party store impl doesn't support this new version's store actions. 

Your point about errors at runtime is a good one, but under the mental model of "store-impl vX supports D4J vY," it's not as relevant. Ideally, we want to detect these things at compile time, but that's not possible here. I'll go ahead and point out that we _could_ prevent this by simply giving the methods default empty() implementations. That would make this approach basically equivalent to the ignore-unknown-actions approach in terms of "compatibility," with the added advantage of being able to statically enumerate the minimal set of actions. But again, I really don't see much to be gained by essentially lying about supporting those actions.

**Comment by Alex1304** [comment]
By "stores supports specific version of D4J", do you mean specific Minor versions (`3.2.*`, `3.3.*`...) or specific Patch versions (`3.2.0`, `3.2.1`...)? Like, if a store impl supports `3.2`, should it support `3.2.0` as well as `3.2.1`, `3.2.2`, etc without the dev having to upgrade store impl along with d4j everytime? Especially if those patch updates are frequent, it could be annoying. But overall I see your point, for Minor versions at least it 100% makes sense. My point is that the following statement:
> Your point about errors at runtime is a good one, but under the mental model of "store-impl vX supports D4J vY," it's not as relevant.

is actually relevant for patch versions upgrades.

**Comment by quanticc** [comment]
Actions are a powerful tool to request specific stuff from a backend, therefore being able to encapsulate a "plan" in a single object.

However mapping core operations 1:1 to actions hides that capability, making it seem equivalent to a method-oriented interface. If we hide the action aspect, how can we give that control to the user?

For example imagine your backend is able to retrieve exactly what you ask, only the names of each guild member. How could a user call this compatible implementation that supports such query?

Maybe we need a hybrid approach: minimal set through method api to enforce compatibility or that "it works", but keep the window open for a `store.execute(action)` method

**Comment by Alex1304** [comment]
First of all I closed the PR in the Stores repo, since we don't know yet if the new api will be in d4j-common or in a new separate repo, I'm putting everything in d4j-common and let you decide later where to move them, that way we can focus on this single PR.

So now I tried to come up with *yet another proposal*. I made `Store` a concrete class, which has one public method `.execute(action)`. To instantiate this class you need to pass a `StoreLayout`, this is the interface that needs to be implemented to support the minimal set of actions required by Discord4J + the ability to define custom ones by overriding a default method returning an `ActionMapper` (by default returns an empty mapper).

I think `Store` can stay in d4j-common, however `StoreLayout` and `*Action` classes could easily be moved to a separate module if necessary, to reduce the amount of dependencies to import for implementors. For now I've put `StoreLayout` and all action classes in a separate subpackage (`discord4j.common.store.layout`).

Feedback requested!!!

**Comment by Alex1304** [comment]
Another update: now I managed to make it so that `StoreLayout` does not depend on `*Action` classes anymore. It means that store implementors will no longer need to know about the existence of those classes and they will be able to remain in the d4j-common module! Now we have both approaches perfectly mixed together: a 100% object-based API for any code that *interacts* with a store (this includes the internals of d4j-core and d4j-gateway), and a 100% method-based API for any code that *implements* a store (so that they know exactly what to implement, while importing the least dependencies possible). This has been made possible thanks to the split between `Store` and `StoreLayout`, so unless I'm forgetting other important points, I'm positive that this is a big step in the right direction.

**Comment by Alex1304** [comment]
Today I have some free time so I wanted to work on javadocs in order to mark this PR as ready for review, but before that I think we should agree about the "incomplete store results" problem. I think it's very important because I believe there are a lot of use cases relying on the accuracy of the data stored (for example if we have a command displaying the guild count of the bot, in the situation where store only contains a partial list of the guilds for whatever reason (eviction, missing intents, network unavailability for distributed stores, etc), there should be a way to make that command display an error message instead of displaying an inaccurate count). Not to mention there's also that spaghetti code to fix in `StoreEntityRetriever#getGuildMembers()`.

I first suggested `PossiblyIncompleteList` but now I see why it is not good, especially since now we have the store return a `Publisher`. I then suggested to add a boolean in the parameter of `Get*s` actions to tell whether we want an accurate list or whatever is available in the store, but I don't feel like changing signature of action constructors and layout methods is a good approach, especially after I realized that the same must also be done for the `Count` actions... So, I tried to think of another alternative, and I think I finally found one that solves most (if not all) of these problems. 

Basically, if a store is unable to provide an accurate list/count of the requested entities, it would error with an `InaccurateResultException`. The trick is that this exception type would carry a `Publisher` that allows to get the result anyway, which gives a lot more flexibility as to how to handle that case :
* If we want to propagate an error, we can do `.onErrorMap(InaccurateResultException.class, e -> ...)`
* If we want to do something else like fallback to an alternate source, we can do `.onErrorResume(InaccurateResultException.class, e -> ...)`
* If we still want to get the result even if it's inaccurate, we can do `.onErrorResume(InaccurateResultException.class, e -> Mono/Flux.from(e.getResultAnyway()))` (method name TBD)

**To be compatible with the current 3.1 behavior, the core module would define a default handler for this new exception, which could be customizable via some new method in `GatewayBootstrap`.** 

I haven't made nor pushed this proposal yet, I would like your feedback first.

**Comment by quanticc** [comment]
Using error signal for this might be stretching it, but the alternative is to build a Result class with some isApproximate method + the data Publisher.. I'm still undecided.

@darichey do you think we should model this use case of stores notifying consumers, telling them they have incomplete information?

**Comment by Alex1304** [comment]
Wouldn't that just be the same as `PossiblyIncompleteList` but with a `Publisher` instead of a `List`? I mean I wouldn't be against that idea since it's close to my initial one, but in comparison to an error signal I think it's a bit less flexible ; namely the bold part in my previous comment wouldn't be possible with a custom return type. It would be cool to be able to let the user freely handle the incomplete case if they want to, by having an option for the EntityRetriever API to just propagate the error instead of silently triggering a requestMembers/returning the inaccurate result.

**Comment by Alex1304** [comment]
Alright I think I'm done with the initial development, I opened the PR for reviews. I added javadocs everywhere, I made static factories for all action classes, and did some other cleanup. For the incomplete result problem I thought it was more reasonable to just drop the subject, so what I did is that I simply duplicated the get members action to have one variant that returns only cached members and the other variant that errors if not all are cached.

Take your time when reviewing this huge diff, if you want some naming changes, source location changes or design improvement suggestions, free to open a bunch of reviews, I'll get through all of them. If you want to start working on a store layout implementation I guess it's also the right moment, anyway I will stay available to respond to any comments :slightly_smiling_face:

**Comment by quanticc** [comment]
Thanks, will review during the weekend. At the same time I'll try to create some implementation, because this PR currently doesn't ship any and we'll need a default in place before merging.

This should highlight some improvements to the API after writing `Store.fromLayout(....)`.

We should consider writing a bridge to bring your own StoreService as this new Store, through a provided dependency to stores-api maybe. At worst the default implementation should be the legacy JDK store pre-installed.

**Comment by Alex1304** [comment]
I think it's indeed better if I let you make the implementation, external point of views will be more efficient at finding stuff to improve in the API than if I just make it myself (although it wouldn't bother me at all, I still have some time to spare). 

Regarding the specifics of that implementation, I guess you will want to go for something very casual at first just so we can at least say "it works", but then if you want to have an implementation that is optimized for write operations (which is kinda the goal of the whole rework), I think a good approach would be the one that @danthonywalker suggested in the Discord server: something where each write operation would be represented as a diff and where read operations would consist of resolving them. A fast-write/slow-read impl would benefit to most Discord bots, maybe to a lesser extent for small ones, but I still think it would be relevant to go for something like this as a default implementation.

**Review by quanticc — CHANGES_REQUESTED** [review]
Thanks again for this PR. I've started working on an implementation that can communicate with legacy stores and found some issues so far.

File: core/src/main/java/discord4j/core/GatewayDiscordClient.java
Comment (Alex1304) [reply core/src/main/java/discord4j/core/GatewayDiscordClient.java:574]: I forgot what's the syntax to add behavior on complete while staying in the same chain. Would .handle or .materialize do it or is there an easier way?

Comment (Alex1304) [reply core/src/main/java/discord4j/core/GatewayDiscordClient.java]: Ok I discovered a bug so I went for a slightly different approach. Basically if the caller of `GatewayDiscordClient#requestMembers` cancels the subscription before completion (for example via a `.take(n)` operator), it will never execute the store action even though the request members payload was sent. So what I did is that as soon as the request payload is sent, it will add the nonce to a `Set` that is shared with the `GatewayBootstrap#acquireConnection` code where the main dispatch() subscription happens. Tell me if that's good.

File: common/src/main/java/discord4j/common/store/action/gateway/GatewayActions.java
Comment (Alex1304) [reply common/src/main/java/discord4j/common/store/action/gateway/GatewayActions.java:241]: When adding the selfId parameter I faced an issue : the `DispatchStoreLayer` does not know the selfId. I could add it as a new param to `DispatchStoreLayer`, but it feels messy. At this point I think we should have a sort of `StoreLayoutFactory` interface that would be in charge of constructing a `StoreLayout` with some meta-information from the bot (like the selfId, maybe the intents and the shard count as well) to allow the implementation to be a minimum aware of the context it is running in.

File: core/src/main/java/discord4j/core/GatewayDiscordClient.java
Comment (quanticc) [reply core/src/main/java/discord4j/core/GatewayDiscordClient.java]: I need some feedback from @darichey about the `completeGuildMembers` action, whether is worth it for us to track if all members were loaded reliably. This is part of the discussion of reviving "fully ready or fully ready with members" futures present in v2.

If we go forward with this and since it adds more complexity, it might be worth exploring applying this to the `asyncMemberChunk` sequence triggered by `GuildDispatchHandlers`. That is also a limit=0 empty query request.

Also your thoughts on the location of the new `instanceof GuildMembersChunk` code, considering `DispatchStoreLayer` also has a special handling for `instanceof GatewayStateChange`. What would be the criteria for placing code under DispatchStoreLayer vs externally?

File: common/src/main/java/discord4j/common/store/action/gateway/GatewayActions.java
Comment (quanticc) [reply common/src/main/java/discord4j/common/store/action/gateway/GatewayActions.java:241]: Seems logical, in fact is what we have currently, StoreService init method uses this data and intents to turn off some stores.

Comment (Alex1304) [reply common/src/main/java/discord4j/common/store/action/gateway/GatewayActions.java:241]: For now we can do it by capturing the selfId from the onReady action so I'm resolving this conversation

File: core/src/main/java/discord4j/core/event/dispatch/ChannelDispatchHandlers.java
Comment (quanticc) [inline core/src/main/java/discord4j/core/event/dispatch/ChannelDispatchHandlers.java:96]: This should be removed as it is unused and causes an error if the channel type is not supported

**Review by quanticc — CHANGES_REQUESTED** [review]

File: core/src/main/java/discord4j/core/event/dispatch/ChannelDispatchHandlers.java
Comment (Alex1304) [reply core/src/main/java/discord4j/core/event/dispatch/ChannelDispatchHandlers.java]: removed

File: core/src/main/java/discord4j/core/shard/GatewayBootstrap.java
Comment (Alex1304) [inline core/src/main/java/discord4j/core/shard/GatewayBootstrap.java:894]: self-reminder to update this line once the default impl is validated

**Review by quanticc — CHANGES_REQUESTED** [review]

File: gateway/src/main/java/discord4j/gateway/state/DispatchStoreLayer.java
Comment (Alex1304) [reply gateway/src/main/java/discord4j/gateway/state/DispatchStoreLayer.java:142]: is it ok like this?

**Review by quanticc — CHANGES_REQUESTED** [review]
The PR changes regarding API changes look good to me.

However I would like to have your implementation proposal / Storage under a new PR because I'm running into some perf issues with it I don't understand yet.

After merging this, I'll provide a temporary default impl using a bridge to old stores-jdk, so we can compare and debug memory concerns more easily.

**Comment by Alex1304** [comment]
yeah i started the testing with my bot, the memory usage is still higher than with 3.1 stores, probably because it's instantiating new maps in each GuildNode/ChannelNode, that turned out to be more costly than expected. However in terms of read/write perfs I've got good results. Whenever I can I'll update the PR to remove the implementation and open a new one later for it, I'll probably rework it from scratch.

**Review by quanticc — APPROVED** [review]
Thanks

**Comment by napstr** [comment]
For anyone else looking at using the snapshots, the quick way to get the old StoreService working with the new Store Api is:

```diff
- .setStoreService(storeService)
+ .setStore(Store.fromLayout(LegacyStoreLayout.of(storeService)))
```

And an example on using the new Store Actions:
```diff
	public Mono<Long> countGuilds() {
-		return this.gatewayDiscordClient.getGatewayResources()
-				.getStateView().getGuildStore().count();
+		return Mono.from(this.gatewayDiscordClient.getGatewayResources()
+				.getStore().execute(ReadActions.countGuilds()));
	}

**Pull request merged**
